src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs

```diff
@@ -229,5 +219,37 @@ internal static bool IsDisabledByProvider(Type providerType, MethodInfo jobFunct
                 return (bool)methodInfo.Invoke(instance, new object[] { jobFunction });
             }
         }
+
+        /// <summary>
+        /// Applies any user registered decorators, followed by any platform decorators.
+        /// See <see cref="WebJobsServiceCollectionExtensions.AddListenerDecorators(Extensions.DependencyInjection.IServiceCollection)"/>.
+        /// </summary>
+        /// <param name="listener">The listener to apply decorators to.</param>
+        /// <param name="functionDefinition">The function the the listener is for.</param>
+        /// <returns>The resulting listener with decorators applied.</returns>
+        private IListener ApplyDecorators(IListener listener, IFunctionDefinition functionDefinition)
+        {
+            Type rootListenerType = listener.GetType();
+            var platformDecorators = _listenerDecorators.Where(p => p.GetType().Assembly == typeof(HostListenerFactory).Assembly);
+            var userDecorators = _listenerDecorators.Except(platformDecorators);
+
+            listener = ApplyDecorators(userDecorators, listener, functionDefinition, rootListenerType);
+
+            // Order is important - platform decorators must be applied AFTER any user decorators, in order.
+            listener = ApplyDecorators(platformDecorators, listener, functionDefinition, rootListenerType);
+
+            return listener;
+        }
+
+        private IListener ApplyDecorators(IEnumerable<IListenerDecorator> decorators, IListener listener, IFunctionDefinition functionDefinition, Type rootListenerType)
+        {
+            foreach (var decorator in decorators)
+            {
+                var context = new ListenerDecoratorContext(functionDefinition, rootListenerType, listener);
+                listener = decorator.Decorate(context);
```

Comment:
This all looks good to me. The only real thought I have is "how do we know if decorators are being applied?" -- from a debugging standpoint.

Could/should we log here that we're applying a decorator? Or add logs in each decorator to say that they're doing something?

I know we only have 2 and they've been built-in for years, but technically this opens some things up...

Reply:
Great suggestion. I'll add a log statement

Reply:
+1 to this.